### PR TITLE
MTM-52577: fix HttpServletRequest autowire issue

### PR DESCRIPTION
**Pull request by chbakar**
Caused by https://github.com/SoftwareAG/cumulocity-clients-java/pull/289

### CODE REVIEW

**Comment by mbak-c8y** [comment]
Stack trace from device-simulator:
com.google.common.util.concurrent.UncheckedExecutionException: javax.ws.rs.ProcessingException: No thread-bound request found: Are you referring to request attributes outside of an actual web request, or processing a request outside of the originally receiving thread? If you are actually operating within a web request and still receive this message, your code is probably running outside of DispatcherServlet: In this case, use RequestContextListener or RequestContextFilter to expose the current request.
    at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2051)
    at com.google.common.cache.LocalCache.get(LocalCache.java:3951)
    at com.google.common.cache.LocalCache$LocalManualCache.get(LocalCache.java:4871)
    at com.cumulocity.microservice.security.token.JwtTokenAuthenticationGuavaCache.get(JwtTokenAuthenticationGuavaCache.java:29)
    at com.cumulocity.microservice.security.token.JwtTokenAuthenticationProvider.authenticate(JwtTokenAuthenticationProvider.java:42)
    at org.springframework.security.authentication.ProviderManager.authenticate(ProviderManager.java:182)
    at com.cumulocity.microservice.security.token.CumulocityOAuthMicroserviceFilter.doFilter(CumulocityOAuthMicroserviceFilter.java:70)
    at org.springframework.security.web.FilterChainProxy$VirtualFilterChain.doFilter(FilterChainProxy.java:346)
    ..
    at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
    at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
    at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
    at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
    at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: javax.ws.rs.ProcessingException: No thread-bound request found: Are you referring to request attributes outside of an actual web request, or processing a request outside of the originally receiving thread? If you are actually operating within a web request and still receive this message, your code is probably running outside of DispatcherServlet: In this case, use RequestContextListener or RequestContextFilter to expose the current request.
    at org.glassfish.jersey.client.ClientRuntime.invoke(ClientRuntime.java:309)
   ...
    at org.glassfish.jersey.client.JerseyInvocation$Builder.get(JerseyInvocation.java:318)
    at com.cumulocity.microservice.security.token.CumulocityCoreAuthenticationClient.getCurrentUser(CumulocityCoreAuthenticationClient.java:95)
    at com.cumulocity.microservice.security.token.CumulocityCoreAuthenticationClient.authenticateUserAndUpdateToken(CumulocityCoreAuthenticationClient.java:36)
    at com.cumulocity.microservice.security.token.JwtTokenAuthenticationProvider$1.call(JwtTokenAuthenticationProvider.java:46)
    at com.cumulocity.microservice.security.token.JwtTokenAuthenticationProvider$1.call(JwtTokenAuthenticationProvider.java:42)
    at com.google.common.cache.LocalCache$LocalManualCache$1.load(LocalCache.java:4876)
    at com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3529)
    at com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2278)
    at com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2155)
    at com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2045)
    ... 44 common frames omitted
Caused by: java.lang.IllegalStateException: No thread-bound request found: Are you referring to request attributes outside of an actual web request, or processing a request outside of the originally receiving thread? If you are actually operating within a web request and still receive this message, your code is probably running outside of DispatcherServlet: In this case, use RequestContextListener or RequestContextFilter to expose the current request.
    at org.springframework.web.context.request.RequestContextHolder.currentRequestAttributes(RequestContextHolder.java:131)

**Review by SAG-jher — DISMISSED** [review]

**Review by mbak-c8y — DISMISSED** [review]

**Review by kowa-c8y — DISMISSED** [review]

**Review by kaczynskid-sag — CHANGES_REQUESTED** [review]
I see no reason to drag the whole request object from `JwtTokenAuthenticationProvider` and through multiple static methods of `CumulocityCoreAuthenticationClient` only to get a single `X-Forwarded-Host` header at the end in `ForwardedHeaderOnRequestFilter`. You can instead just use the `RequestContextHolder` directly in `ForwardedHeaderOnRequestFilter`, or in the method where it's instance is created instead.

Second issue is that added `RequestContextListener` is designed to be used in `web.xml` as servlet listener, while Spring Boot already defines `RequestContextFilter` in `WebMvcAutoConfiguration` that is handling setting and clearing of `RequestAttributes` in `RequestContextHolder`. So adding this is most likely redundant or at least should be made conditional.

Additionally `CumulocityCoreAuthenticationClient` instead of being a utility class with only static methods used in a single, anonymous implementation of `JwtTokenAuthenticationLoader`, could instead be a bean that implements that interface, with `Environment` as a dependency to get the `baseUrl`, especially that this dependency is not used for anything else in `JwtTokenAuthenticationProvider`.

Method `com.cumulocity.microservice.security.token.CumulocityOAuthMicroserviceFilter#readCredentials()` should not throw checked `ParseException` as i breaks the abstraction level of the caller `doFilter` method. Instead `JWTParser#parse` should be wrapped in a private method that handles `ParseException` and converts it into runtime `AuthenticationServiceException`.

**Comment by chbakar** [comment]
> I see no reason to drag the whole request object from `JwtTokenAuthenticationProvider` and through multiple static methods of `CumulocityCoreAuthenticationClient` only to get a single `X-Forwarded-Host` header at the end in `ForwardedHeaderOnRequestFilter`. You can instead just use the `RequestContextHolder` directly in `ForwardedHeaderOnRequestFilter`, or in the method where it's instance is created instead.
> 
> Second issue is that added `RequestContextListener` is designed to be used in `web.xml` as servlet listener, while Spring Boot already defines `RequestContextFilter` in `WebMvcAutoConfiguration` that is handling setting and clearing of `RequestAttributes` in `RequestContextHolder`. So adding this is most likely redundant or at least should be made conditional.
> 
> Additionally `CumulocityCoreAuthenticationClient` instead of being a utility class with only static methods used in a single, anonymous implementation of `JwtTokenAuthenticationLoader`, could instead be a bean that implements that interface, with `Environment` as a dependency to get the `baseUrl`, especially that this dependency is not used for anything else in `JwtTokenAuthenticationProvider`.
> 
> Method `com.cumulocity.microservice.security.token.CumulocityOAuthMicroserviceFilter#readCredentials()` should not throw checked `ParseException` as i breaks the abstraction level of the caller `doFilter` method. Instead `JWTParser#parse` should be wrapped in a private method that handles `ParseException` and converts it into runtime `AuthenticationServiceException`.

All good points. I will address them in the next commits. As for your third point, I think It should be outside of the scope of this.
As an no-issue task following this one

File: microservice/security/src/main/java/com/cumulocity/microservice/security/token/CumulocityOAuthConfiguration.java [hunk 1]
```diff
@@ -40,6 +42,12 @@ public JwtTokenAuthenticationProvider jwtTokenAuthenticationProvider(
         return new JwtTokenAuthenticationProvider(environment, jwtAuthenticatedTokenCache);
     }
 
+    @Bean
+    @ConditionalOnMissingBean
```
Comment (chbakar) [inline microservice/security/src/main/java/com/cumulocity/microservice/security/token/CumulocityOAuthConfiguration.java:46]: `RequestContextListener` may be redundant, so loading it only when not already in the context. Also it seems It is not always found, so initializing it here

**Review by mbak-c8y — DISMISSED** [review]

**Review by kowa-c8y — DISMISSED** [review]

**Review by kowa-c8y — DISMISSED** [review]

**Review by mbak-c8y — DISMISSED** [review]

**Review by kaczynskid-sag — DISMISSED** [review]

**Review by kowa-c8y — DISMISSED** [review]

**Review by c8y-kh4mell — DISMISSED** [review]

**Review by mbak-c8y — DISMISSED** [review]

**Comment by chbakar** [comment]
Since, this is a blocker. I please request someone who can to force merge this. 
I know from @kaczynskid-sag there is issue with dev-x, but I think we should still merge this.

@c8y-kh4mell @SAG-jher

**Review by mbak-c8y — APPROVED** [review]

**Review by kowa-c8y — APPROVED** [review]

**Pull request merged**